src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp

```diff
@@ -391,10 +392,74 @@ namespace OpenLoco::World::MapGenerator
         call(0x004597FD, regs);
     }
 
+    // 0x0042E731
+    static void generateMiscBuildingType0(const BuildingObject* buildingObj, const size_t id)
+    {
+        registers regs;
+        regs.ebp = X86Pointer(buildingObj);
+        regs.ebx = id;
+        call(0x0042E731, regs);
+    }
+
+    // 0x0042E893
+    static void generateMiscBuildingType1(const BuildingObject* buildingObj, const size_t id)
+    {
+        registers regs;
+        regs.ebp = X86Pointer(buildingObj);
+        regs.ebx = id;
+        call(0x0042E893, regs);
+    }
+
+    // 0x0042EA29
+    static void generateMiscBuildingType2(const BuildingObject* buildingObj, const size_t id)
+    {
+        registers regs;
+        regs.ebp = X86Pointer(buildingObj);
+        regs.ebx = id;
+        call(0x0042EA29, regs);
+    }
+
+    // 0x0042EB94
+    static void generateMiscBuildingType3(const BuildingObject* buildingObj, const size_t id)
+    {
+        registers regs;
+        regs.ebp = X86Pointer(buildingObj);
+        regs.ebx = id;
+        call(0x0042EB94, regs);
+    }
+
     // 0x0042E6F2
     static void generateMiscBuildings()
     {
-        call(0x0042E6F2);
+        GameCommands::setUpdatingCompanyId(CompanyId::neutral);
+
+        for (auto id = 0U; id < ObjectManager::getMaxObjects(ObjectType::building); id++)
+        {
+            auto* buildingObj = ObjectManager::get<BuildingObject>(id);
+            if (buildingObj == nullptr)
+            {
+                continue;
+            }
+
+            if (!buildingObj->hasFlags(BuildingObjectFlags::miscBuilding))
+            {
+                continue;
+            }
+
+            if (buildingObj->hasFlags(BuildingObjectFlags::isHeadquarters))
+            {
+                continue;
+            }
+
+            static std::array<std::function<void(const BuildingObject*, const size_t)>, 4> generatorFunctions = {
+                generateMiscBuildingType0,
+                generateMiscBuildingType1,
+                generateMiscBuildingType2,
+                generateMiscBuildingType3,
+            };
+
+            generatorFunctions[buildingObj->generatorFunction](buildingObj, id);
```

Comment:
are you sure these are all the generator functions? I looked in all the building objects and the exhaustive list of generator values is `[0, 1, 2, 3, 4, 6, 7, 8, 12]`

Reply:
oh but all the buildings with generatorFunction >3 have one of those flags set, so it doesn't need a generator function. interesting why they wouldn't just be set to a single value like 0xFF or something

Reply:
These are the four functions considered for these 'miscellaneous' buildings, yes. You're right that there are more functions involved, though. Could you verify these do not occur when the relevant misc flag is not set?

Reply:
sorry i edited my comment a couple times, but i did investigate and all have the `miscBuilding` flag set when generator < 4 so this code is good